File: sdks/typescript/test/levers.test.ts
// Tests for the SPEC §17.5–§17.8 levers introduced in alpha.7.

import { test } from "node:test";
import assert from "node:assert/strict";

import {
  PROTOCOL_VERSION,
  SCOPE_MEETING_ATTEND,
  bundleHash,
  generateAgent,
  generateChallenge,
  generateHumanRoot,
  issueDelegation,
  issuePolicyVerdict,
  issueVerificationReceipt,
  receiptHash,
  signChallenge,
  verifierContextHash,
  verifyBundle,
  verifyPolicyVerdict,
  verifyVerificationReceipt,
  type Anchor,
  type AnchorResolver,
  type Constraint,
  type ConstraintEvaluator,
  type DelegationCert,
  type ProofBundle,
  type VerifierContext,
  type VerifyResult,
} from "../src/index.js";

async function goodBundle(): Promise<{ bundle: ProofBundle; certID: string; humanID: string }> {
  const { root, privateKey: rootPriv } = await generateHumanRoot();
  const { agent, privateKey: agentPriv } = await generateAgent("L Bot", "custom");
  const now = Math.floor(Date.now() / 1000);
  const cert: DelegationCert = {
    cert_id: "lever-cert",
    version: PROTOCOL_VERSION,
    issuer_id: root.id,
    issuer_pub_key: root.public_key,
    subject_id: agent.id,
    subject_pub_key: agent.public_key,
    scope: [SCOPE_MEETING_ATTEND],
    issued_at: now,
    expires_at: now + 86400,
    signature: { ed25519: new Uint8Array(0), ml_dsa_65: new Uint8Array(0) },
  };
  await issueDelegation(cert, rootPriv);
  const challenge = generateChallenge();
  const sig = await signChallenge(challenge, now, agentPriv);
  return {
    bundle: {
      agent_id: agent.id,
      agent_pub_key: agent.public_key,
      delegations: [cert],
      challenge,
      challenge_at: now,
      challenge_sig: sig,
    },
    certID: cert.cert_id,
    humanID: root.id,
  };
}

// ---------------------------------------------------------------------------
// Lever 1: VerificationReceipt
// ---------------------------------------------------------------------------

test("verification receipt round-trip", async () => {
  const { bundle } = await goodBundle();
  const { agent: v, privateKey: vPriv } = await generateAgent("v", "verifier");
  const result = await verifyBundle(bundle, {});
  const r = await issueVerificationReceipt(
    bundle, result, v.id, v.public_key, vPriv, null, Math.floor(Date.now() / 1000),
  );
  const err = await verifyVerificationReceipt(r);
  assert.equal(err, null);
  assert.equal(r.decision, "authorized_agent");
});

test("verification receipt detects tampering", async () => {
  const { bundle } = await goodBundle();
  const { agent: v, privateKey: vPriv } = await generateAgent("v", "verifier");
  const result = await verifyBundle(bundle, {});
  const r = await issueVerificationReceipt(
    bundle, result, v.id, v.public_key, vPriv, null, Math.floor(Date.now() / 1000),
  );
  r.decision = "revoked";
  const err = await verifyVerificationReceipt(r);
  assert.notEqual(err, null);
});

test("verification receipt detects bundle substitution", async () => {
  const { bundle: b1 } = await goodBundle();
  const { bundle: b2 } = await goodBundle();
  const { agent: v, privateKey: vPriv } = await generateAgent("v", "verifier");
  const result = await verifyBundle(b1, {});
  const r = await issueVerificationReceipt(
    b1, result, v.id, v.public_key, vPriv, null, Math.floor(Date.now() / 1000),
  );
  r.bundle_hash = bundleHash(b2);
  const err = await verifyVerificationReceipt(r);
  assert.notEqual(err, null);
});

test("verification receipt chain linkage", async () => {
  const { bundle } = await goodBundle();
  const { agent: v, privateKey: vPriv } = await generateAgent("v", "verifier");
  const result = await verifyBundle(bundle, {});
  const r1 = await issueVerificationReceipt(
    bundle, result, v.id, v.public_key, vPriv, null, Math.floor(Date.now() / 1000),
  );
  const prev = receiptHash(r1);
  const r2 = await issueVerificationReceipt(
    bundle, result, v.id, v.public_key, vPriv, prev, Math.floor(Date.now() / 1000),
  );
  assert.deepEqual(r2.prev_hash, prev);
  // Tampering r1 changes its hash — chain pointer in r2 would no longer match.
  r1.decision = "tampered";
  const prevAfter = receiptHash(r1);
  assert.notDeepEqual(prev, prevAfter);
});

test("bundleHash deterministic", async () => {
  const { bundle } = await goodBundle();
  assert.deepEqual(bundleHash(bundle), bundleHash(bundle));
  assert.equal(bundleHash(bundle).length, 32);
});

// ---------------------------------------------------------------------------
// Lever 2: PolicyVerdict
// ---------------------------------------------------------------------------

const SECRET = new Uint8Array(32).fill(0x33);

test("policy verdict round-trip", () => {
  const now = Math.floor(Date.now() / 1000);
  const ctx = verifierContextHash({});
  const v = issuePolicyVerdict("vid", "agent-A", "meeting:attend", true, ctx, now, now + 3600, SECRET);
  assert.equal(verifyPolicyVerdict(v, SECRET, "agent-A", "meeting:attend", ctx, now), null);
});

test("policy verdict deny returns policy_verdict_denied", () => {
  const now = Math.floor(Date.now() / 1000);
  const ctx = verifierContextHash({});
  const v = issuePolicyVerdict("v", "a", "s", false, ctx, now, now + 3600, SECRET);
  const err = verifyPolicyVerdict(v, SECRET, "a", "s", ctx, now);
  assert.match(err ?? "", /policy_verdict_denied/);
});

test("policy verdict wrong secret rejected", () => {
  const now = Math.floor(Date.now() / 1000);
  const ctx = verifierContextHash({});
  const v = issuePolicyVerdict("v", "a", "s", true, ctx, now, now + 3600, SECRET);
  const wrong = new Uint8Array(32).fill(0x44);
  assert.notEqual(verifyPolicyVerdict(v, wrong, "a", "s", ctx, now), null);
});

test("policy verdict context_hash mismatch", () => {
  const now = Math.floor(Date.now() / 1000);
  const ctxA = verifierContextHash({ current_lat: 37, current_lon: -122 });
  const ctxB = verifierContextHash({ current_lat: 51.5, current_lon: -0.1 });
  const v = issuePolicyVerdict("v", "a", "s", true, ctxA, now, now + 3600, SECRET);
  assert.notEqual(verifyPolicyVerdict(v, SECRET, "a", "s", ctxB, now), null);
});

test("policy verdict expired", () => {
  const now = Math.floor(Date.now() / 1000);
  const ctx = verifierContextHash({});
  const v = issuePolicyVerdict("v", "a", "s", true, ctx, now - 7200, now - 3600, SECRET);
  assert.notEqual(verifyPolicyVerdict(v, SECRET, "a", "s", ctx, now), null);
});

test("policy verdict fast-path skips live policy", async () => {
  const { bundle } = await goodBundle();
  const now = Math.floor(Date.now() / 1000);
  const ctx = verifierContextHash({});
  const verdict = issuePolicyVerdict(
    "vid", bundle.agent_id, "meeting:attend", true, ctx, now - 60, now + 3600, SECRET,
  );
  let liveCalls = 0;
  const res = await verifyBundle(bundle, {
    required_scope: "meeting:attend",
    policy: {
      async evaluatePolicy() {
        liveCalls += 1;
        return false; // would deny
      },
    },
    policy_verdict: verdict,
    policy_secret: SECRET,
  });
  assert.equal(res.valid, true, res.error_reason);
  assert.equal(liveCalls, 0);
});

test("policy verdict fast-path: cached deny", async () => {
  const { bundle } = await goodBundle();
  const now = Math.floor(Date.now() / 1000);
  const ctx = verifierContextHash({});
  const verdict = issuePolicyVerdict(
    "vid", bundle.agent_id, "meeting:attend", false, ctx, now - 60, now + 3600, SECRET,
  );
  let liveCalls = 0;
  const res = await verifyBundle(bundle, {
    required_scope: "meeting:attend",
    policy: { async evaluatePolicy() { liveCalls += 1; return true; } },
    policy_verdict: verdict,
    policy_secret: SECRET,
  });
  assert.equal(res.valid, false);
  assert.equal(res.identity_status, "scope_denied");
  assert.equal(liveCalls, 0);
});

test("policy verdict falls back to live policy when stale", async () => {
  const { bundle } = await goodBundle();
  const now = Math.floor(Date.now() / 1000);
  const ctx = verifierContextHash({});
  const expired = issuePolicyVerdict(
    "vid", bundle.agent_id, "meeting:attend", true, ctx, now - 7200, now - 3600, SECRET,
  );
  let liveCalls = 0;
  const res = await verifyBundle(bundle, {
    required_scope: "meeting:attend",
    policy: { async evaluatePolicy() { liveCalls += 1; return true; } },
    policy_verdict: expired,
    policy_secret: SECRET,
  });
  assert.equal(res.valid, true, res.error_reason);
  assert.equal(liveCalls, 1);
});

// ---------------------------------------------------------------------------
// Lever 3: ConstraintEvaluator
// ---------------------------------------------------------------------------

async function bundleWithCustomConstraint(type: string): Promise<ProofBundle> {
  const { root, privateKey: rootPriv } = await generateHumanRoot();
  const { agent, privateKey: agentPriv } = await generateAgent("C Bot", "custom");
  const now = Math.floor(Date.now() / 1000);
  const cert: DelegationCert = {
    cert_id: "custom-cert",
    version: PROTOCOL_VERSION,
    issuer_id: root.id,
    issuer_pub_key: root.public_key,
    subject_id: agent.id,
    subject_pub_key: agent.public_key,
    scope: [SCOPE_MEETING_ATTEND],
    constraints: [{ type } as Constraint],
    issued_at: now,
    expires_at: now + 3600,
    signature: { ed25519: new Uint8Array(0), ml_dsa_65: new Uint8Array(0) },
  };
  await issueDelegation(cert, rootPriv);
  const challenge = generateChallenge();
  const sig = await signChallenge(challenge, now, agentPriv);
  return {
    agent_id: agent.id,
    agent_pub_key: agent.public_key,
    delegations: [cert],
    challenge,
    challenge_at: now,
    challenge_sig: sig,
  };
}

test("constraint evaluator: unknown type fails closed", async () => {
  const bundle = await bundleWithCustomConstraint("verify.max_concurrent_sessions");
  const res = await verifyBundle(bundle, {});
  assert.equal(res.valid, false);
  assert.equal(res.identity_status, "constraint_unknown");
});

test("constraint evaluator: registered allow", async () => {
  const bundle = await bundleWithCustomConstraint("verify.max_concurrent_sessions");
  const ev: ConstraintEvaluator = { async evaluate() { return true; } };
  const res = await verifyBundle(bundle, {
    constraint_evaluators: { "verify.max_concurrent_sessions": ev },
  });
  assert.equal(res.valid, true, res.error_reason);
});

test("constraint evaluator: registered deny", async () => {
  const bundle = await bundleWithCustomConstraint("verify.max_concurrent_sessions");
  const ev: ConstraintEvaluator = { async evaluate() { return false; } };
  const res = await verifyBundle(bundle, {
    constraint_evaluators: { "verify.max_concurrent_sessions": ev },
  });
  assert.equal(res.valid, false);
  assert.equal(res.identity_status, "constraint_denied");
});

test("constraint evaluator: unverifiable routes correctly", async () => {
  const bundle = await bundleWithCustomConstraint("verify.needs_context");
  const ev: ConstraintEvaluator = { async evaluate() { return "unverifiable"; } };
  const res = await verifyBundle(bundle, {
    constraint_evaluators: { "verify.needs_context": ev },
  });
  assert.equal(res.valid, false);
  assert.equal(res.identity_status, "constraint_unverifiable");
});

// ---------------------------------------------------------------------------
// Lever 4: AnchorResolver
// ---------------------------------------------------------------------------

test("anchor resolver populates result", async () => {
  const { bundle, humanID } = await goodBundle();
  const anchor: Anchor = {
    type: "enterprise_sso",
    provider: "okta",
    reference: "opaque",
    verified_at: 1000,
  };
  const resolver: AnchorResolver = {
    async resolveAnchor(id) {
      return id === humanID ? anchor : null;
    },
  };
  const res = await verifyBundle(bundle, { anchor_resolver: resolver });
  assert.equal(res.valid, true, res.error_reason);
  assert.equal(res.anchor?.provider, "okta");
});

test("anchor resolver error is non-fatal", async () => {
  const { bundle } = await goodBundle();
  const resolver: AnchorResolver = {
    async resolveAnchor() { throw new Error("directory down"); },
  };
  const res = await verifyBundle(bundle, { anchor_resolver: resolver });
  assert.equal(res.valid, true, res.error_reason);
  assert.equal(res.anchor, undefined);
});

// Cross-SDK byte-equivalence is covered by cross_sdk.test.ts which loads
// the canonical fixture file testvectors/v1/cross_sdk_vectors.json.

test("audit observes anchor", async () => {
  const { bundle, humanID } = await goodBundle();
  const anchor: Anchor = {
    type: "email", provider: "google", reference: "h:abc", verified_at: 100,
  };
  const logged: VerifyResult[] = [];
  const res = await verifyBundle(bundle, {
    anchor_resolver: { async resolveAnchor(id) { return id === humanID ? anchor : null; } },
    audit: { async logVerification(r) { logged.push(r); } },
  });
  assert.equal(res.valid, true);
  assert.equal(logged.length, 1);
  assert.equal(logged[0]!.anchor?.provider, "google");
});

File: levers_test.go
package ratify_test

import (
	"bytes"
	"crypto/sha256"
	"errors"
	"strings"
	"testing"
	"time"

	. "github.com/identities-ai/ratify-protocol"
)

// ---------------------------------------------------------------------------
// Lever 1: VerificationReceipt — SPEC §17.5
// ---------------------------------------------------------------------------
//
// A signed attestation that "this verifier, with this key, saw this exact
// bundle, and reached this decision, at this time." The cryptographic
// complement of AuditProvider: an audit provider can throw entries away,
// but a chain of VerificationReceipts is unforgeable.

func TestVerificationReceipt_RoundTrip(t *testing.T) {
	bundle, _ := providerBundle(t)
	verifier, verifierPriv, _ := GenerateAgentKeypair("verifier-1", "verifier")
	result := Verify(bundle, VerifyOptions{})
	if !result.Valid {
		t.Fatalf("setup: bundle should verify; got %s", result.ErrorReason)
	}

	r, err := IssueVerificationReceipt(
		bundle, result,
		verifier.ID, verifier.PublicKey, verifierPriv,
		nil, // genesis
		time.Now().Unix(),
	)
	if err != nil {
		t.Fatalf("IssueVerificationReceipt: %v", err)
	}
	if err := VerifyVerificationReceipt(r); err != nil {
		t.Fatalf("VerifyVerificationReceipt: %v", err)
	}
	if r.Decision != "authorized_agent" {
		t.Errorf("Decision = %q, want authorized_agent", r.Decision)
	}
	if len(r.BundleHash) != sha256.Size {
		t.Errorf("BundleHash should be 32 bytes; got %d", len(r.BundleHash))
	}
	zeros := make([]byte, sha256.Size)
	if !bytes.Equal(r.PrevHash, zeros) {
		t.Errorf("genesis prev_hash should be all zeros")
	}
}

func TestVerificationReceipt_DetectsTampering(t *testing.T) {
	bundle, _ := providerBundle(t)
	verifier, verifierPriv, _ := GenerateAgentKeypair("verifier-1", "verifier")
	result := Verify(bundle, VerifyOptions{})

	r, _ := IssueVerificationReceipt(
		bundle, result, verifier.ID, verifier.PublicKey, verifierPriv,
		nil, time.Now().Unix(),
	)
	// Flip the decision after signing.
	r.Decision = "revoked"
	if err := VerifyVerificationReceipt(r); err == nil {
		t.Fatal("tampered receipt MUST not verify")
	}
}

func TestVerificationReceipt_DetectsBundleSubstitution(t *testing.T) {
	bundle1, _ := providerBundle(t)
	bundle2, _ := providerBundle(t)
	verifier, verifierPriv, _ := GenerateAgentKeypair("v", "verifier")
	result := Verify(bundle1, VerifyOptions{})

	r, _ := IssueVerificationReceipt(
		bundle1, result, verifier.ID, verifier.PublicKey, verifierPriv,
		nil, time.Now().Unix(),
	)
	// Replace BundleHash with that of a different bundle — sig must fail.
	r.BundleHash, _ = BundleHash(bundle2)
	if err := VerifyVerificationReceipt(r); err == nil {
		t.Fatal("substituted bundle_hash MUST invalidate receipt")
	}
}

func TestVerificationReceipt_ChainLinkage(t *testing.T) {
	bundle, _ := providerBundle(t)
	verifier, verifierPriv, _ := GenerateAgentKeypair("v", "verifier")
	result := Verify(bundle, VerifyOptions{})

	r1, _ := IssueVerificationReceipt(
		bundle, result, verifier.ID, verifier.PublicKey, verifierPriv,
		nil, time.Now().Unix(),
	)
	prev, err := ReceiptHash(r1)
	if err != nil {
		t.Fatalf("ReceiptHash: %v", err)
	}
	r2, _ := IssueVerificationReceipt(
		bundle, result, verifier.ID, verifier.PublicKey, verifierPriv,
		prev, time.Now().Unix(),
	)
	// r2.PrevHash must match the digest of r1's signable bytes.
	if !bytes.Equal(r2.PrevHash, prev) {
		t.Errorf("r2.PrevHash should chain off r1")
	}
	// Forge a chain break: tamper r1 retroactively, recompute its hash → r2's
	// chain pointer is now wrong.
	r1.Decision = "tampered"
	prevAfterTamper, _ := ReceiptHash(r1)
	if bytes.Equal(prev, prevAfterTamper) {
		t.Fatal("ReceiptHash must change when receipt is tampered")
	}
}

func TestBundleHash_Deterministic(t *testing.T) {
	bundle, _ := providerBundle(t)
	a, _ := BundleHash(bundle)
	b, _ := BundleHash(bundle)
	if !bytes.Equal(a, b) {
		t.Error("BundleHash must be deterministic")
	}
	if len(a) != sha256.Size {
		t.Errorf("BundleHash must be 32 bytes; got %d", len(a))
	}
}

// ---------------------------------------------------------------------------
// Lever 2: PolicyVerdict — SPEC §17.6
// ---------------------------------------------------------------------------

func TestPolicyVerdict_RoundTrip(t *testing.T) {
	secret := bytes.Repeat([]byte{0x33}, 32)
	now := time.Now()
	ctxHash, _ := VerifierContextHash(VerifierContext{})

	v, err := IssuePolicyVerdict(
		"verdict-1", "agent-A", "meeting:attend", true, ctxHash,
		now.Unix(), now.Add(time.Hour).Unix(), secret,
	)
	if err != nil {
		t.Fatalf("IssuePolicyVerdict: %v", err)
	}
	if err := VerifyPolicyVerdict(v, secret, "agent-A", "meeting:attend", ctxHash, now); err != nil {
		t.Fatalf("VerifyPolicyVerdict: %v", err)
	}
}

func TestPolicyVerdict_DenyVerdict(t *testing.T) {
	secret := bytes.Repeat([]byte{0x33}, 32)
	now := time.Now()
	ctxHash, _ := VerifierContextHash(VerifierContext{})

	v, _ := IssuePolicyVerdict(
		"v", "agent-A", "payments:send", false, ctxHash,
		now.Unix(), now.Add(time.Hour).Unix(), secret,
	)
	err := VerifyPolicyVerdict(v, secret, "agent-A", "payments:send", ctxHash, now)
	if err == nil || !strings.Contains(err.Error(), "policy_verdict_denied") {
		t.Errorf("explicit deny should return policy_verdict_denied; got %v", err)
	}
}

func TestPolicyVerdict_WrongSecretRejected(t *testing.T) {
	secret := bytes.Repeat([]byte{0x33}, 32)
	wrongSecret := bytes.Repeat([]byte{0x44}, 32)
	now := time.Now()
	ctxHash, _ := VerifierContextHash(VerifierContext{})

	v, _ := IssuePolicyVerdict("v", "a", "s", true, ctxHash, now.Unix(), now.Add(time.Hour).Unix(), secret)
	if err := VerifyPolicyVerdict(v, wrongSecret, "a", "s", ctxHash, now); err == nil {
		t.Fatal("verdict signed by another secret must not verify")
	}
}

func TestPolicyVerdict_ContextHashMismatch(t *testing.T) {
	secret := bytes.Repeat([]byte{0x33}, 32)
	now := time.Now()
	ctxA, _ := VerifierContextHash(VerifierContext{HasLocation: true, CurrentLat: 37.0, CurrentLon: -122.0})
	ctxB, _ := VerifierContextHash(VerifierContext{HasLocation: true, CurrentLat: 51.5, CurrentLon: -0.1})

	v, _ := IssuePolicyVerdict("v", "a", "s", true, ctxA, now.Unix(), now.Add(time.Hour).Unix(), secret)
	// Caller is in a different context (London vs SF) — verdict must reject.
	if err := VerifyPolicyVerdict(v, secret, "a", "s", ctxB, now); err == nil {
		t.Fatal("verdict for one context must not apply to another")
	}
}

func TestPolicyVerdict_Expired(t *testing.T) {
	secret := bytes.Repeat([]byte{0x33}, 32)
	now := time.Now()
	ctxHash, _ := VerifierContextHash(VerifierContext{})

	v, _ := IssuePolicyVerdict("v", "a", "s", true, ctxHash, now.Add(-2*time.Hour).Unix(), now.Add(-time.Hour).Unix(), secret)
	if err := VerifyPolicyVerdict(v, secret, "a", "s", ctxHash, now); err == nil {
		t.Fatal("expired verdict must not verify")
	}
}

func TestPolicyVerdict_FastPathSkipsLivePolicy(t *testing.T) {
	bundle, _ := providerBundle(t)
	secret := bytes.Repeat([]byte{0x33}, 32)
	now := time.Now()
	ctxHash, _ := VerifierContextHash(VerifierContext{})

	v, _ := IssuePolicyVerdict(
		"vid", bundle.AgentID, "meeting:attend", true, ctxHash,
		now.Add(-time.Minute).Unix(), now.Add(time.Hour).Unix(),
		secret,
	)
	livePolicy := &fakePolicy{allow: false} // would fail if consulted

	res := Verify(bundle, VerifyOptions{
		RequiredScope: "meeting:attend",
		Policy:        livePolicy,
		PolicyVerdict: v,
		PolicySecret:  secret,
	})
	if !res.Valid {
		t.Fatalf("verdict fast-path should accept: %s", res.ErrorReason)
	}
	if livePolicy.calls != 0 {
		t.Errorf("live policy MUST NOT be called when valid verdict is present; calls=%d", livePolicy.calls)
	}
}

func TestPolicyVerdict_FallsBackOnStaleVerdict(t *testing.T) {
	bundle, _ := providerBundle(t)
	secret := bytes.Repeat([]byte{0x33}, 32)
	now := time.Now()
	ctxHash, _ := VerifierContextHash(VerifierContext{})

	// Expired verdict — verdict-verify fails; should fall back to live policy.
	v, _ := IssuePolicyVerdict(
		"vid", bundle.AgentID, "meeting:attend", true, ctxHash,
		now.Add(-2*time.Hour).Unix(), now.Add(-time.Hour).Unix(),
		secret,
	)
	livePolicy := &fakePolicy{allow: true}

	res := Verify(bundle, VerifyOptions{
		RequiredScope: "meeting:attend",
		Policy:        livePolicy,
		PolicyVerdict: v,
		PolicySecret:  secret,
	})
	if !res.Valid {
		t.Fatalf("expected valid (live policy allows): %s", res.ErrorReason)
	}
	if livePolicy.calls != 1 {
		t.Errorf("live policy should be consulted when verdict stale; calls=%d", livePolicy.calls)
	}
}

func TestPolicyVerdict_CachedDenyFastPath(t *testing.T) {
	bundle, _ := providerBundle(t)
	secret := bytes.Repeat([]byte{0x33}, 32)
	now := time.Now()
	ctxHash, _ := VerifierContextHash(VerifierContext{})

	v, _ := IssuePolicyVerdict(
		"vid", bundle.AgentID, "meeting:attend", false, ctxHash,
		now.Add(-time.Minute).Unix(), now.Add(time.Hour).Unix(),
		secret,
	)
	livePolicy := &fakePolicy{allow: true}

	res := Verify(bundle, VerifyOptions{
		RequiredScope: "meeting:attend",
		Policy:        livePolicy,
		PolicyVerdict: v,
		PolicySecret:  secret,
	})
	if res.Valid {
		t.Fatal("cached deny must reject")
	}
	if res.IdentityStatus != "scope_denied" {
		t.Errorf("IdentityStatus = %q, want scope_denied", res.IdentityStatus)
	}
	if livePolicy.calls != 0 {
		t.Errorf("live policy MUST NOT be called when valid cached deny exists; calls=%d", livePolicy.calls)
	}
}

// ---------------------------------------------------------------------------
// Lever 3: ConstraintEvaluator registry — SPEC §17.7
// ---------------------------------------------------------------------------

type fakeConcurrentSessionsEvaluator struct {
	max   int
	count int
}

func (f *fakeConcurrentSessionsEvaluator) Evaluate(
	c Constraint, _ string, _ VerifierContext, _ time.Time,
) error {
	if f.count > f.max {
		return errors.New("too many concurrent sessions")
	}
	return nil
}

func providerBundleWithCustomConstraint(t *testing.T, constraintType string) (*ProofBundle, string) {
	t.Helper()
	humanRoot, humanPriv, _ := GenerateHumanRootKeypair()
	agent, agentPriv, _ := GenerateAgentKeypair("Custom Bot", "custom")
	now := time.Now()
	cert := &DelegationCert{
		CertID:        "custom-constraint-cert",
		Version:       ProtocolVersion,
		IssuerID:      humanRoot.ID,
		IssuerPubKey:  humanRoot.PublicKey,
		SubjectID:     agent.ID,
		SubjectPubKey: agent.PublicKey,
		Scope:         []string{ScopeMeetingAttend},
		IssuedAt:      now.Unix(),
		ExpiresAt:     now.Add(time.Hour).Unix(),
		Constraints: []Constraint{
			{Type: ConstraintType(constraintType)},
		},
	}
	if err := IssueDelegation(cert, humanPriv); err != nil {
		t.Fatalf("IssueDelegation: %v", err)
	}
	challenge, _ := GenerateChallenge()
	sig, _ := SignChallenge(challenge, now.Unix(), agentPriv)
	return &ProofBundle{
		AgentID: agent.ID, AgentPubKey: agent.PublicKey,
		Delegations:  []DelegationCert{*cert},
		Challenge:    challenge,
		ChallengeAt:  now.Unix(),
		ChallengeSig: sig,
	}, cert.CertID
}

func TestConstraintEvaluator_UnknownTypeFailsClosed(t *testing.T) {
	bundle, _ := providerBundleWithCustomConstraint(t, "verify.max_concurrent_sessions")
	res := Verify(bundle, VerifyOptions{})
	if res.Valid {
		t.Fatal("unknown constraint type MUST fail closed")
	}
	if res.IdentityStatus != "constraint_unknown" {
		t.Errorf("IdentityStatus = %q, want constraint_unknown", res.IdentityStatus)
	}
}

func TestConstraintEvaluator_RegistryAllow(t *testing.T) {
	bundle, _ := providerBundleWithCustomConstraint(t, "verify.max_concurrent_sessions")
	ev := &fakeConcurrentSessionsEvaluator{max: 10, count: 3}

	res := Verify(bundle, VerifyOptions{
		ConstraintEvaluators: map[string]ConstraintEvaluator{
			"verify.max_concurrent_sessions": ev,
		},
	})
	if !res.Valid {
		t.Fatalf("registered evaluator should allow; got %s: %s", res.IdentityStatus, res.ErrorReason)
	}
}

func TestConstraintEvaluator_RegistryDeny(t *testing.T) {
	bundle, _ := providerBundleWithCustomConstraint(t, "verify.max_concurrent_sessions")
	ev := &fakeConcurrentSessionsEvaluator{max: 10, count: 999}

	res := Verify(bundle, VerifyOptions{
		ConstraintEvaluators: map[string]ConstraintEvaluator{
			"verify.max_concurrent_sessions": ev,
		},
	})
	if res.Valid {
		t.Fatal("registered evaluator returned error; should deny")
	}
	if res.IdentityStatus != "constraint_denied" {
		t.Errorf("IdentityStatus = %q, want constraint_denied", res.IdentityStatus)
	}
}

type fakeUnverifiableEvaluator struct{}

func (fakeUnverifiableEvaluator) Evaluate(
	_ Constraint, _ string, _ VerifierContext, _ time.Time,
) error {
	return ErrConstraintUnverifiable
}

func TestConstraintEvaluator_UnverifiableRoutes(t *testing.T) {
	bundle, _ := providerBundleWithCustomConstraint(t, "verify.needs_context")
	res := Verify(bundle, VerifyOptions{
		ConstraintEvaluators: map[string]ConstraintEvaluator{
			"verify.needs_context": fakeUnverifiableEvaluator{},
		},
	})
	if res.Valid {
		t.Fatal("unverifiable extension constraint must fail")
	}
	if res.IdentityStatus != "constraint_unverifiable" {
		t.Errorf("IdentityStatus = %q, want constraint_unverifiable", res.IdentityStatus)
	}
}

// ---------------------------------------------------------------------------
// Lever 4: AnchorResolver / anchor-bound audit — SPEC §17.8
// ---------------------------------------------------------------------------

type fakeAnchorResolver struct {
	anchors map[string]*Anchor
	err     error
}

func (f *fakeAnchorResolver) ResolveAnchor(humanID string) (*Anchor, error) {
	if f.err != nil {
		return nil, f.err
	}
	return f.anchors[humanID], nil
}

func TestAnchorResolver_PopulatesResult(t *testing.T) {
	bundle, _ := providerBundle(t)
	// Bundle has no anchor pre-attached; we resolve from human_id.
	humanID := bundle.Delegations[len(bundle.Delegations)-1].IssuerID
	anchor := &Anchor{Type: "enterprise_sso", Provider: "okta", Reference: "opaque-ref", VerifiedAt: 1000}

	res := Verify(bundle, VerifyOptions{
		AnchorResolver: &fakeAnchorResolver{anchors: map[string]*Anchor{humanID: anchor}},
	})
	if !res.Valid {
		t.Fatalf("expected valid: %s", res.ErrorReason)
	}
	if res.Anchor == nil {
		t.Fatal("Anchor should be populated")
	}
	if res.Anchor.Provider != "okta" {
		t.Errorf("Anchor.Provider = %q, want okta", res.Anchor.Provider)
	}
}

func TestAnchorResolver_ResolverErrorIsNonFatal(t *testing.T) {
	bundle, _ := providerBundle(t)
	res := Verify(bundle, VerifyOptions{
		AnchorResolver: &fakeAnchorResolver{err: errors.New("identity directory down")},
	})
	if !res.Valid {
		t.Fatalf("anchor resolver error MUST NOT fail the bundle; got %s", res.ErrorReason)
	}
	if res.Anchor != nil {
		t.Error("Anchor should be nil when resolver errors")
	}
}

func TestAnchorResolver_NoResolverLeavesAnchorNil(t *testing.T) {
	bundle, _ := providerBundle(t)
	res := Verify(bundle, VerifyOptions{})
	if !res.Valid {
		t.Fatalf("expected valid: %s", res.ErrorReason)
	}
	if res.Anchor != nil {
		t.Error("Anchor should be nil when no resolver is configured")
	}
}

func TestAnchorResolver_AuditObservesAnchor(t *testing.T) {
	bundle, _ := providerBundle(t)
	humanID := bundle.Delegations[len(bundle.Delegations)-1].IssuerID
	anchor := &Anchor{Type: "email", Provider: "google", Reference: "h:abc", VerifiedAt: 100}
	audit := &fakeAudit{}

	res := Verify(bundle, VerifyOptions{
		AnchorResolver: &fakeAnchorResolver{anchors: map[string]*Anchor{humanID: anchor}},
		Audit:          audit,
	})
	if !res.Valid {
		t.Fatalf("expected valid: %s", res.ErrorReason)
	}
	if len(audit.results) != 1 {
		t.Fatalf("expected 1 audit entry; got %d", len(audit.results))
	}
	if audit.results[0].Anchor == nil || audit.results[0].Anchor.Provider != "google" {
		t.Errorf("audit entry should observe anchor; got %+v", audit.results[0].Anchor)
	}
}

// ---------------------------------------------------------------------------
// Lever 5: IsRevoked closure remains functional (deprecation is doc-only)
// ---------------------------------------------------------------------------

func TestLegacyIsRevoked_StillWorks(t *testing.T) {
	bundle, certID := providerBundle(t)
	res := Verify(bundle, VerifyOptions{
		IsRevoked: func(c string) bool { return c == certID },
	})
	if res.Valid {
		t.Fatal("legacy closure should still revoke")
	}
	if res.IdentityStatus != "revoked" {
		t.Errorf("IdentityStatus = %q, want revoked", res.IdentityStatus)
	}
}
